source/loader/layers/sanitizer/asan_interceptor.cpp

```diff
@@ -207,85 +251,93 @@ bool SanitizerInterceptor::preLaunchKernel(ur_kernel_handle_t Kernel,
             "{} of size {} at kernel <{}> LID({}, {}, {}) GID({}, "
             "{}, {})",
-            AH->IsWrite ? "WRITE" : "READ", AH->AccessSize, Func, AH->LID0,
-            AH->LID1, AH->LID2, AH->GID0, AH->GID1, AH->GID2);
+            AH->IsWrite ? "WRITE" : "READ", AH->AccessSize, KernelName.c_str(),
+            AH->LID0, AH->LID1, AH->LID2, AH->GID0, AH->GID1, AH->GID2);
         context.logger.always("  #0 {} {}:{}", Func, File, AH->Line);
         if (!AH->IsRecover) {
-            abort();
+            exit(1);
         }
     }
 }
 
-std::string SanitizerInterceptor::getKernelName(ur_kernel_handle_t Kernel) {
-    size_t KernelNameSize = 0;
-    [[maybe_unused]] auto Res = context.urDdiTable.Kernel.pfnGetInfo(
-        Kernel, UR_KERNEL_INFO_FUNCTION_NAME, 0, nullptr, &KernelNameSize);
-    assert(Res == UR_RESULT_SUCCESS);
-
-    std::vector<char> KernelNameBuf(KernelNameSize);
-    Res = context.urDdiTable.Kernel.pfnGetInfo(
-        Kernel, UR_KERNEL_INFO_FUNCTION_NAME, KernelNameSize,
-        KernelNameBuf.data(), nullptr);
-    assert(Res == UR_RESULT_SUCCESS);
-
-    return std::string(KernelNameBuf.data(), KernelNameSize - 1);
-}
-
 ur_result_t SanitizerInterceptor::allocShadowMemory(
     ur_context_handle_t Context, std::shared_ptr<DeviceInfo> &DeviceInfo) {
     if (DeviceInfo->Type == DeviceType::CPU) {
-        context.logger.error("Unsupport device type");
-        return UR_RESULT_ERROR_INVALID_ARGUMENT;
+        if (!m_IsInASanContext) {
+            static std::once_flag OnceFlag;
+            bool Result = true;
+            std::call_once(OnceFlag, [&]() {
+                Result = m_ShadowMemInited = SetupShadowMem();
+            });
+
+            if (!Result) {
+                context.logger.error("Failed to allocate shadow memory");
+                return UR_RESULT_ERROR_OUT_OF_RESOURCES;
+            }
+        }
+
+        DeviceInfo->ShadowOffset = LOW_SHADOW_BEGIN;
+        DeviceInfo->ShadowOffsetEnd = HIGH_SHADOW_END;
     } else if (DeviceInfo->Type == DeviceType::GPU_PVC) {
         /// SHADOW MEMORY MAPPING (PVC, with CPU 47bit)
         ///   Host/Shared USM : 0x0              ~ 0x0fff_ffff_ffff
         ///   ?               : 0x1000_0000_0000 ~ 0x1fff_ffff_ffff
         ///   Device USM      : 0x2000_0000_0000 ~ 0x3fff_ffff_ffff
         constexpr size_t SHADOW_SIZE = 1ULL << 46;
-
-        // TODO: Protect Bad Zone
-        auto Result = context.urDdiTable.VirtualMem.pfnReserve(
-            Context, nullptr, SHADOW_SIZE, (void **)&DeviceInfo->ShadowOffset);
-        if (Result != UR_RESULT_SUCCESS) {
-            context.logger.error("Failed to allocate shadow memory on PVC: {}",
-                                 Result);
-            return Result;
+        // FIXME: Currently, Level-Zero doesn't create independent VAs for each contexts,
+        // which will cause out-of-resource error when users use multiple contexts
+        static uptr ShadowOffset, ShadowOffsetEnd;
```

Comment:
why use static here?

Reply:
As mentioned on comment, Level-Zero use one VA for multiple contexts, so if user creates multiple contexts, we have to reserve multiple huge shadow memories, which will cause out-of-resource error. So I added "static" here to let one shadow memory shared with all contexts.

Reply:
can call_once be used instead? it might be more intuitive

Reply:
This piece of code will also suffer multi thread issue (CPU as well). I'm going to redesign DeviceInfo in next patch, and let each ur_device_handle_t correspond to only one unique instance of it, so that there will be only one shadow memory offset in each DeviceInfo.

For now, let's just leave it as this.

Reply:
do you mean this 'static' code will be removed after your redesign so that the multi-thread issue will be resolved?

Reply:
Yep, because DeviceInfo will be initialized only once.

Reply:
looks good to me

Reply:
So this is race now, right? Once all this is resolved, might be worth adding multi-threaded tests that exercise this code path.

Reply:
> So this is a race now, right? Once all this is resolved, might be worth adding multi-threaded tests that exercise this code path.

Yes, I will.